build.rs

```diff
@@ -49,6 +54,166 @@ fn unpack<P: AsRef<Path>>(tgz: impl Read, outdir: P) -> Result<(), std::io::Erro
     Ok(())
 }
 
+// We need the interior mutability of `RefCell` to work around the fact that the
+// `func_macro` callback is not called with a `mut` reference. We need
+// `AssertUnwindSafe` to allow `RefCell` to be used within a `catch_unwind`
+// context inside `bindgen`. We use `Rc` to so that we can get back the `T`
+// after `bindgen` is done with the `Callbacks`, by keeping the reference alive
+// without having to make it `&'static` instead.
+type HideUnwinding<T> = Rc<AssertUnwindSafe<RefCell<T>>>;
+
+struct Callbacks {
+    wrapped: bindgen::CargoCallbacks,
+    state: HideUnwinding<BTreeMap<String, Vec<u8>>>,
+}
+
+impl std::fmt::Debug for Callbacks {
+    fn fmt(&self, fmt: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
+        fmt.debug_struct("Callbacks")
+            .field("wrapped", &self.wrapped)
+            .field("state", &self.state.borrow())
+            .finish()
+    }
+}
+
+impl Callbacks {
+    fn new(bt: HideUnwinding<BTreeMap<String, Vec<u8>>>) -> Self {
+        let wrapped = bindgen::CargoCallbacks;
+        let state = Rc::clone(&bt);
+        Self { wrapped, state }
+    }
+}
+
+impl bindgen::callbacks::ParseCallbacks for Callbacks {
+    fn func_macro(&self, name: &str, value: &[&[u8]]) {
+        self.state.borrow_mut().insert(name.to_owned(), value.join(" ".as_bytes()));
+    }
+
+    fn include_file(&self, filename: &str) {
+        self.wrapped.include_file(filename)
+    }
+}
+
+fn generate(pairs: impl IntoIterator<Item=(String,String)>)
```

Comment:
I am not at all happy with how complex this function is. I hope to refactor it, or at least document it well, during part 2.

Reply:
(I might even get to it in part 1, here.)